Hunting Queries/SecurityEvent/ProcessEntropy.yaml

```diff
@@ -16,77 +16,112 @@ tactics:
-      | where EventID == 4688
-      // excluding well known processes
-      | where NewProcessName !endswith ':\\Windows\\System32\\conhost.exe' and ParentProcessName !endswith ':\\Windows\\System32\\conhost.exe'
-      | where ParentProcessName !endswith ":\\Windows\\System32\\wuauclt.exe" and NewProcessName !startswith "C:\\Windows\\SoftwareDistribution\\Download\\Install\\AM_Delta_Patch_"
-      | where ParentProcessName !has ":\\WindowsAzure\\GuestAgent_" and NewProcessName !has ":\\WindowsAzure\\GuestAgent_"
-      | where ParentProcessName !has ":\\WindowsAzure\\WindowsAzureNetAgent_" and NewProcessName !has ":\\WindowsAzure\\WindowsAzureNetAgent_"
-      | where ParentProcessName !has ":\\ProgramData\\Microsoft\\Windows Defender\\platform\\" and ParentProcessName !endswith "\\MpCmdRun.exe" 
-      | project TimeGenerated, EventID, Computer, SubjectUserSid, Account, AccountType, Process, NewProcessName, CommandLine, ParentProcessName
-  ) on Computer, Process
-  | project TimeGenerated, EventID, Computer, SubjectUserSid, Account, Weight, AdjustedProcessEntropy, FullDecimalProcessEntropy = ProcessEntropy, Process, NewProcessName, CommandLine, ParentProcessName, TotalProcessCountOnHost, ProcessCountOnHost, DistinctComputersWithProcessCount
-  | sort by Weight asc, AdjustedProcessEntropy asc, NewProcessName asc
+  ) on Computer 
+  | distinct Computer, Process, AllHostsProcessCount 
+  //Getting a decimal value for later computation 
+  | extend AHPCValue = todecimal(AllHostsProcessCount));
+  // Need the count of each class in my bucket or also said as count of ProcName(Class) per Host(Bucket) for use in the entropy calc 
+  let PCoH = Include 
+  | summarize ProcessCountOnHost = count() by Computer, Process 
+  | join kind=inner ( 
+  Include 
+  ) on Computer,Process 
+  | distinct Computer, Process, ProcessCountOnHost 
+  //Getting a decimal value for later computation 
+  | extend PCoHValue = todecimal(ProcessCountOnHost); 
+  let Combined = DCwPC 
+  | join ( 
+  AHPC 
+  ) on Computer, Process 
+  | join ( 
+  PCoH 
+  ) on Computer, Process;
+  let Results = Combined 
+  // Entropy calculation 
+  | extend ProcessEntropy = -log2(PCoHValue/AHPCValue)*(PCoHValue/AHPCValue) 
+  // Calculating Weight, see details in description 
+  | extend Weight = toreal(ProcessEntropy*ProcessCountOnHost*DistinctHostsProcessCount) 
+  // Remove or increase value to see processes with low entropy, meaning more common. 
+  | where Weight <= 100
+  | project Computer, Process, Weight , ProcessEntropy, AllHostsProcessCount, ProcessCountOnHost, DistinctHostsProcessCount; 
+  // Join back full entry 
+  Results 
+  | join kind= inner ( 
+  SecEvents
+  | project TimeGenerated, EventID, Computer, SubjectUserSid, Account, AccountType, Process, NewProcessName, CommandLine, ParentProcessName, _ResourceId, SourceComputerId 
+  ) on Computer, Process 
+  | project TimeGenerated, EventID, Computer, SubjectUserSid, Account, AccountType, Weight, ProcessEntropy,  
```

Comment:
Is there a way to group final results (specially  timestamp, commandline) so that final output has 1 row per rare process without loosing context.

Sample output showed multiple rows (40-50) for single process  (e.g. MOMPerfSnapshotHelper.exe,  MicrosoftEdgeUpdate.exe, WmiApSrv.exe, TiWorker.exe, gpupdate.exe) with same weight and parent but different timestamp and same commandline or commadline with minor cardinality.

Reply:
Done